### ISSUE TITLE
 'NoneType' object has no attribute 'before'

### ISSUE THREAD

**Issue by fizyk** [Schwarzschild/Elephants#1]
I've mode some decorator based on yours to add a callback ability whenever I get results from cache, but for some time, I'm getting this weird error:

``` python
    File "...../elephants/strongdict.py", line 142, in __getitem__
        self.__move_to_beginning(l)
    File "...../elephants/strongdict.py", line 114, in __move_to_beginning
        top.before = l
    exceptions.AttributeError: 'NoneType' object has no attribute 'before'
```

The decorator code looks like this:

``` python
class memorize(object):

    def __init__(self, cache_callback=None, tlimit=None, limit=None):
        '''param callable cache_callback: callback to call whenever some
            additional work is required on successful cache retrieval.
            Should accept two arguments: original function, and current key
        :param int tlimit: time limit elements will be kept in cache if set
        :param int limit: total number of items that will be kept in cache
            if set
        '''
        self.cache_callback = cache_callback
        self.cache = StrongDict(tlimit=tlimit, limit=limit)

    def __call__(self, f):
        '''
        :param callable f: wrapped function
        '''
        f.cache = self.cache

        @wraps(f)  # Makes the wrap function think it has the name of f.
        def wrapped_f(*args, **kwargs):

            k = (args, frozenset(kwargs.items()))

            try:
                results = self.cache[k]
                if self.cache_callback:
                    self.cache_callback(self.cache, k)
                return results
            except KeyError:

                results = f(*args, **kwargs)
                self.cache[k] = results
                return results

        return wrapped_f
```

**Comment by Schwarzschild** [comment]
It is a nice idea to have the call back.
I have not been seeing this message.
Can you send me a usage test example for your memorize?

Thank you.

**Comment by fizyk** [comment]
It's same as yours, except If I want to know the length of acache I just run len(function.cache).

We are using it on method used by some twisted code to cache calculations based on various input  *args, cache might get large as we rarely clear it (it's not that often we push a new data to the process), but will be accessed rather frequently to check if value is cached or not.

Can it be that some cache elements gets garbage collected?

For now we've replaced StrongDict with ordinary dict and see how it will perform.

**Comment by Schwarzschild** [comment]
I'm sorry, I don't mean to be lazy.  Can you provide specific test code you are using so I can just cut/paste and run in my debugger?

Thank you.

**Comment by fizyk** [comment]
I'm sorry, I didn't mean any disrespect, it's just it's fairly simple and tests only if it is pinned to StrongDict correctly. Might be missing a lot.

``` python

@memorize()
def memorized_method(*args):
    return len(args)

def test_memorize():

    assert len(memorized_method.cache) == 0

    first_call = (1, 2, 3, 4)
    second_call = (2, 3, 4)

    memorized_method(*first_call)

    assert len(memorized_method.cache) == 1

    # original structure of cache keys
    assert (first_call, frozenset()) in memorized_method.cache.keys()

    memorized_method(*first_call)
    memorized_method(*first_call)

    assert len(memorized_method.cache) == 1

    memorized_method(*second_call)

    assert len(memorized_method.cache) == 2

    memorized_method.cache.clear()

    assert len(memorized_method.cache) == 0
```

**Comment by Schwarzschild** [comment]
No, worries, I wasn't offended in anyway.  I just wanted test code.

I put all your code in a file called memorize.py and ran it with nosetests, as follows, and had no errors at all.  I cannot reproduce your errors.  I also reran nosetests on test_strongdict.py and all was OK.

ms@ny:~/py/elephants_support$ nosetests memorize.py                              

## .                                                                                

Ran 1 test in 0.000s                                                             

OK  
ms@ny:~/py/elephants_support$

**Comment by fizyk** [comment]
Test case rus fine for us too. It's just when the program runs (really long), we get this error from time to time.

**Comment by Schwarzschild** [comment]
That is interesting.  I'd like to observe that problem too.  How long is "really long"?  How many elements are there at the point errors start?  Are there any deletes from the cache before you see the error messages.  The best thing would be to have a failing example.  Could you provide a loop that causes the problem?

**Comment by fizyk** [comment]
Hi, sorry for my late response. How long is really long... few days probably Let assume it's day, and at the moment it's called at minimum 10 times a sec with different args. Lets assume at least a day, that would give us 846 000 cached items.
Sorry, but I don't have the exact numbers.
Our assumption was that the item dissapeared in gc somewhere, and since replacing this with standard dict, I haven't seen any new exception.

**Comment by Schwarzschild** [comment]
No need to apologize, I like the feedback and want to help resolve your problem and make Elephants better.

However, I have not been able to reproduce the problem you are having.  It would be very helpful if you could provide some usage code that demonstrates the problem.

Thank you.

**Comment by Schwarzschild** [comment]
I finally saw this error too.  I believe it is s fixed now in the new 1.0.0 version posted tonight.  I was able to reproduce it after calling the memoized method with the 'clear_elephants_cache' special argument.

M.
